lib/requests/request_executor.rb

```diff
@@ -0,0 +1,319 @@
+    end
+
+    def execute(command)
+      chosen_node = nil
+      chosen_node_index = -1
+
+      await_first_topology_update
+      selector = @_node_selector
+      chosen_node = selector.current_node
+      chosen_node_index = selector.current_node_index
+
+      begin
+        response = execute_command(command, chosen_node)
+      rescue TopologyNodeDownException
+        response = handle_server_down(command, chosen_node, chosen_node_index)
+      end
+
+      response
+    end  
+
+    protected
+    def is_first_topology_update_tries_expired?
+      @_first_topology_updates_tries >= MaxFirstTopologyUpdatesTries
+    end
+
+    def await_first_topology_update()
+      is_fulfilled = false
+      first_topology_update = @_first_topology_update
+
+      if @_without_topology
+        return;
+      end
+
+      @_await_first_topology_lock.synchronize do
+        if first_topology_update.equal?(@_first_topology_update)
+          is_fulfilled = true == first_topology_update;
+
+          if false == first_topology_update
+            start_first_topology_update(@_last_known_urls)
+          end  
+        end
+      end  
+
+      if !is_fulfilled
+        if is_first_topology_update_tries_expired?
+          raise DatabaseLoadFailureException, 'Max topology update tries reached'
+        elsif
+          sleep 0.1
+          await_first_topology_update
+        end    
+      end      
+    end
+
+    def prepare_command(command, server_node)
+      command.create_request(server_node)
+      request = command.to_request_options
+
+      @headers.each do |header, value|
+        request.add_field(header, value)
+      end
+
+      if !@_without_topology
+        request.add_field("Topology-Etag", @_topology_etag)
+      end
+
+      request
+    end
+
+    def execute_command(command, server_node)      
+      if !command.is_a?(RavenCommand)
+        raise InvalidOperationException, 'Not a valid command'
+      end
+
+      if !server_node.is_a?(ServerNode)
+        raise InvalidOperationException, 'Not a valid server node'
+      end
+
+      request = prepare_command(command, server_node)
+      
+      begin
+        response = http_client(server_node).request(request)
+      rescue
+        raise TopologyNodeDownException, "Node #{server_node.url} is down"
+      end    
+
+      is_server_error = [Net::HTTPRequestTimeOut, Net::HTTPBadGateway,
+        Net::HTTPGatewayTimeOut, Net::HTTPServiceUnavailable].any? { 
+        |response_type| response.is_a?(response_type)
+      }
+
+      if response.is_a?(Net::HTTPNotFound)
+        response.body = nil
+      end  
+
+      if is_server_error
+        if command.was_failed?
+          message = 'Unsuccessfull request'
+          json = response.json
+
+          if json && json.Error
+              message += ": #{json.Error}";
+          end
+
+          raise UnsuccessfulRequestException, message
+        end
+
+        raise TopologyNodeDownException, "Node #{server_node.url} is down"
```

Comment:
Because of what?

Reply:
if response has one of the following statuses [Net::HTTPRequestTimeOut, Net::HTTPBadGateway,
       Net::HTTPGatewayTimeOut, Net::HTTPServiceUnavailable] it means that node is down
this part of logic was taken from Python client

Reply:
The problem is we'd like to know which of the errors you listed caused it
exactly.

Reply:
@gregolsky Net::HTTPRequestTimeOut, Net::HTTPBadGateway,
Net::HTTPGatewayTimeOut, Net::HTTPServiceUnavailable means that http client didn't connected to server due to the timeout or http server was down. So there were NO json response with "Error" property describing raven exception

Reply:
I can just add a text description like "HTTP 408: Request timeout" but nothing more

Reply:
Yeah, info on which one of `Net::HTTPRequestTimeOut, Net::HTTPBadGateway,
Net::HTTPGatewayTimeOut, Net::HTTPServiceUnavailable` is it.